File: backend/app/api/v1/endpoints/exec.py
"""Exec endpoints: task CRUD, stop/retry, stats, log pagination (after_seq replay)."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Query

from app.api.deps import DbDep, UserDep
from app.core.response import Result
from app import schemas
from app.services import exec_service

router = APIRouter(prefix="/exec", tags=["exec"])


@router.get("/tasks", response_model=Result)
def list_tasks(
    db: DbDep,
    user: UserDep,
    task_no: str | None = None,
    name: str | None = None,
    status: str | None = None,
    kind: str | None = None,
    start: str | None = None,
    end: str | None = None,
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=100)] = 10,
):
    return Result.ok(exec_service.list_tasks(db, user, task_no, name, status, kind, start, end, page, size))


@router.post("/tasks", response_model=Result)
def create_task(db: DbDep, user: UserDep, data: schemas.ExecTaskCreate):
    return Result.ok(exec_service.create_task(db, user, data))


@router.get("/tasks/{task_id}", response_model=Result)
def get_task(db: DbDep, user: UserDep, task_id: int):
    return Result.ok(exec_service.get_task(db, user, task_id))


@router.get("/tasks/{task_id}/stats", response_model=Result)
def task_stats(db: DbDep, user: UserDep, task_id: int):
    return Result.ok(exec_service.task_stats(db, user, task_id))


@router.post("/tasks/{task_id}/stop", response_model=Result)
def stop_task(db: DbDep, user: UserDep, task_id: int):
    return Result.ok(exec_service.stop_task(db, user, task_id))


@router.post("/tasks/{task_id}/retry", response_model=Result)
def retry_task(db: DbDep, user: UserDep, task_id: int):
    return Result.ok(exec_service.retry_task(db, user, task_id))


@router.get("/tasks/{task_id}/hosts/{task_host_id}/logs", response_model=Result)
def task_logs(
    db: DbDep,
    user: UserDep,
    task_id: int,
    task_host_id: int,
    after_seq: int = 0,
    size: Annotated[int, Query(ge=1, le=1000)] = 200,
):
    return Result.ok(exec_service.get_logs(db, user, task_id, task_host_id, after_seq, size))


@router.get("/tasks/{task_id}/hosts/{task_host_id}/ws-token", response_model=Result)
def task_ws_token(db: DbDep, user: UserDep, task_id: int, task_host_id: int):
    return Result.ok(exec_service.ws_token(db, user, task_id, task_host_id))

File: frontend/src/api/exec.ts
/**
 * Exec API — new file frontend/src/api/exec.ts (stage 3 batch 1)
 * Contract: api-design v2.1 §5 (/exec/tasks), shapes @ backend schemas e04d85e.
 * Realtime log WS: GET ws-token -> WS /ws/exec/{task_host_id}?token= (5min JWT bound
 * to task_host_id; close 4401=token, 4404=not found). REST remains the replay source
 * via after_seq cursor; stop is REST-only.
 */
import http from './http'
import type {
  Result,
  Page,
  ExecTaskOut,
  ExecTaskDetail,
  ExecTaskCreate,
  ExecTaskQuery,
  ExecStats,
  ExecLogPage,
  WsTokenOut,
} from './types'

export async function getTasks(params?: ExecTaskQuery): Promise<Page<ExecTaskOut>> {
  const { data } = await http.get<Result<Page<ExecTaskOut>>>('/exec/tasks', { params })
  return data.data
}

export async function createTask(payload: ExecTaskCreate): Promise<ExecTaskDetail> {
  const { data } = await http.post<Result<ExecTaskDetail>>('/exec/tasks', payload)
  return data.data
}

export async function getTask(taskId: number): Promise<ExecTaskDetail> {
  const { data } = await http.get<Result<ExecTaskDetail>>(`/exec/tasks/${taskId}`)
  return data.data
}

export async function getTaskStats(taskId: number): Promise<ExecStats> {
  const { data } = await http.get<Result<ExecStats>>(`/exec/tasks/${taskId}/stats`)
  return data.data
}

export async function stopTask(taskId: number): Promise<void> {
  await http.post<Result<void>>(`/exec/tasks/${taskId}/stop`)
}

export async function retryTask(taskId: number): Promise<void> {
  await http.post<Result<void>>(`/exec/tasks/${taskId}/retry`)
}

export async function getLogs(
  taskId: number,
  taskHostId: number,
  afterSeq = 0,
  size = 200,
): Promise<ExecLogPage> {
  const { data } = await http.get<Result<ExecLogPage>>(
    `/exec/tasks/${taskId}/hosts/${taskHostId}/logs`,
    { params: { after_seq: afterSeq, size } },
  )
  return data.data
}

export async function getWsToken(taskId: number, taskHostId: number): Promise<string> {
  const { data } = await http.get<Result<WsTokenOut>>(
    `/exec/tasks/${taskId}/hosts/${taskHostId}/ws-token`,
  )
  return data.data.token
}
